controller/client/client.go

```diff
@@ -49,33 +57,116 @@ func (c *Client) Close() error {
 }
 
 func (c *Client) ListTasks(ctx context.Context) ([]*tasks.Task, error) {
-	reader, _, err := c.request(ctx, "GET", "/tasks", nil)
+	resp, err := c.request(ctx, "GET", "/tasks", nil)
 	if err != nil {
 		return nil, err
 	}
 
 	var res []*tasks.Task
-	err = json.NewDecoder(reader).Decode(&res)
+	err = json.NewDecoder(resp.Body).Decode(&res)
 	if err != nil {
 		return nil, err
 	}
 
 	return res, nil
 }
 
-func (c *Client) UpdateTask(ctx context.Context, r *UpdateTaskRequest) (io.ReadCloser, int, error) {
+func (c *Client) UpdateTask(ctx context.Context, UUID string, r *UpdateTaskRequest) (*tasks.Task, error) {
 	var body bytes.Buffer
 	err := json.NewEncoder(&body).Encode(r)
 	if err != nil {
-		return nil, -1, err
+		return nil, err
+	}
+
+	resp, err := c.request(ctx, "PATCH", "/tasks/"+UUID, bytes.NewReader(body.Bytes()))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, ErrRequestFailed{resp.StatusCode}
+	}
+
+	var res *tasks.Task
+	err = json.NewDecoder(resp.Body).Decode(&res)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+func (c *Client) GetTask(ctx context.Context, UUID string) (*tasks.Task, error) {
+	resp, err := c.request(ctx, "GET", "/tasks/"+UUID, nil)
+	if err != nil {
+		return nil, err
 	}
 
-	return c.request(ctx, "PUT", "/task", bytes.NewReader(body.Bytes()))
+	if resp.StatusCode != http.StatusOK {
+		return nil, ErrRequestFailed{resp.StatusCode}
+	}
+
+	var res *tasks.Task
+	err = json.NewDecoder(resp.Body).Decode(&res)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
 }
 
-func (c *Client) request(ctx context.Context, method string, path string, body io.Reader, headers ...string) (io.ReadCloser, int, error) {
+func (c *Client) CreateStorageTask(ctx context.Context, storageTask *tasks.StorageTask) (*tasks.Task, error) {
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(storageTask)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.request(ctx, "POST", "/tasks/storage", bytes.NewReader(body.Bytes()))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusCreated {
+		return nil, ErrRequestFailed{resp.StatusCode}
+	}
+
+	var res *tasks.Task
+	err = json.NewDecoder(resp.Body).Decode(&res)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+func (c *Client) CreateRetrievalTask(ctx context.Context, retrievalTask *tasks.RetrievalTask) (*tasks.Task, error) {
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(retrievalTask)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.request(ctx, "POST", "/tasks/retrieval", bytes.NewReader(body.Bytes()))
```

Comment:
I realised I posted this in @willscott's PR, because I had not noticed the PRs were atop each other.

You can use the `body` buffer directly in these :)